synapse/storage/databases/state/store.py

```diff
@@ -237,45 +257,179 @@ def _get_state_for_group_using_cache(
-
         member_filter, non_member_filter = state_filter.get_member_split()
 
         # Now we look them up in the member and non-member caches
         (
             non_member_state,
             incomplete_groups_nm,
         ) = self._get_state_for_groups_using_cache(
-            groups, self._state_group_cache, state_filter=non_member_filter
+            state_groups, self._state_group_cache, state_filter=non_member_filter
         )
 
-        (member_state, incomplete_groups_m,) = self._get_state_for_groups_using_cache(
-            groups, self._state_group_members_cache, state_filter=member_filter
+        (member_state, incomplete_groups_m) = self._get_state_for_groups_using_cache(
+            state_groups, self._state_group_members_cache, state_filter=member_filter
         )
 
         state = dict(non_member_state)
-        for group in groups:
-            state[group].update(member_state[group])
-
-        # Now fetch any missing groups from the database
+        for state_group in state_groups:
+            state[state_group].update(member_state[state_group])
 
+        # We may have only got one or none of the events for the group so mark those as
+        # incomplete that need fetching from the database.
         incomplete_groups = incomplete_groups_m | incomplete_groups_nm
 
+        return (state, incomplete_groups)
+
+    @cancellable
+    @trace
+    @tag_args
+    async def _get_state_for_client_filtering(
+        self, state_group_ids: Iterable[int], user_id_viewing_events: str
+    ) -> Dict[int, MutableStateMap[str]]:
+        """Get a state map for each state group ID provided that is necessary to filter
+        the corresponding events before being displayed to clients from the perspective
+        of the `user_id_viewing_events`.
+
+        Args:
+            state_group_ids: The state groups to fetch
+            user_id_viewing_events: User ID that will be viewing the events that correspond
+                to the state groups
+
+        Returns:
+            Dict of state_group ID to state map.
+        """
+
+        def _get_state_for_client_filtering_txn(
+            txn: LoggingTransaction, groups: Iterable[int]
+        ) -> Mapping[int, MutableStateMap[str]]:
+
+            sql = """
+                WITH RECURSIVE sgs(state_group) AS (
+                    VALUES(CAST(? AS bigint))
+                    UNION ALL
+                    SELECT prev_state_group FROM state_group_edges e, sgs s
+                    WHERE s.state_group = e.state_group
+                )
+                SELECT
+                    type, state_key, event_id
+                FROM state_groups_state
+                WHERE
+                    state_group IN (
+                        SELECT state_group FROM sgs
+                    )
+                    AND (type = ? AND state_key = ?)
+                ORDER BY
+                    type,
+                    state_key,
```

Comment:
I think we can safely get rid of the `type` and `state_key` ordering since we always filter to one kind in the `where` now